File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.kt
class Solution {
    fun solution(numbers: IntArray, target: Int): Int {
        var answer = 0
        
        fun dfs(index:Int, temp:Int){
            if(index < numbers.size-1)
            {
                dfs(index+1, temp+numbers[index])
                dfs(index+1, temp-numbers[index])
            }
            else 
            {
                if(temp+numbers[index] == target || temp-numbers[index] == target){
                    answer++
                }
            }
        }
        dfs(0,0)
        return answer
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.js
let answer = 0;

function solution(numbers, target) {
    dfs(numbers, target, 0, 0);
    return answer;
}

function dfs(numbers, target, sum, index) {
    // 탈출 조건: 모든 숫자를 사용한 경우
    if (index === numbers.length) {
        if (sum === target) {
            answer++;
        }
        return;
    }

    // 현재 숫자를 더하거나 빼는 두 갈래로 탐색
    dfs(numbers, target, sum + numbers[index], index + 1);
    dfs(numbers, target, sum - numbers[index], index + 1);
}
